File: designPatterns/SOLID_design_principles/OpenClosedPrinciple/javascript/passOpenClosedPrinciple.js
// Polymorphic open-closed principle
class Question {
  constructor(description) {
    this.description = description;
  }

  printQuestionChoices() {
    console.log("");
  }
}

class BooleanQuestion extends Question {
  constructor(description) {
    super(description);
  }

  printQuestionChoices() {
    console.log("1. True");
    console.log("2. False");
  }
}

class MultipleChoiceQuestion extends Question {
  constructor(description, options) {
    super(description);
    this.options = options;
  }

  printQuestionChoices() {
    this.options.forEach((option, index) => {
      console.log(`${index + 1}. ${option}`);
    });
  }
}

class TextQuestion extends Question {
  constructor(description) {
    super(description);
  }

  printQuestionChoices() {
    console.log("Answer: _______________");
  }
}

class RangeQuestion extends Question {
  constructor(description) {
    super(description);
  }

  printQuestionChoices() {
    console.log("Minimum: ______________");
    console.log("Maximum: ______________");
  }
}

// this is open to be extended
// closed for modification as we do not need to touch this function
function printQuiz(questions) {
  questions.forEach(question => {
    console.log(question.description);
    question.printQuestionChoices();
  });
}

// only need to extend new features here
const questions = [
  new BooleanQuestion("This video is useful."),
  new MultipleChoiceQuestion("What is your favorite language?", [
    "CSS",
    "HTML",
    "JS",
    "Python",
  ]),
  new TextQuestion("Describe your favorite JS feature."),
  new RangeQuestion("What is the speed limit in your city?"),
];

printQuiz(questions);

File: interviewPrep/design_principles/SOLID/OpenClosedPrinciple/python/passOpenClosedPrinciple.py
# Polymorphic open-closed principle
class Question:
    def __init__(self, description):
        self.description = description

    def print_question_choices(self):
        print("")


class BooleanQuestion(Question):
    def __init__(self, description):
        super().__init__(description)

    def print_question_choices(self):
        print("1. True")
        print("2. False")


class MultipleChoiceQuestion(Question):
    def __init__(self, description, options):
        super().__init__(description)
        self.options = options

    def print_question_choices(self):
        for index, option in enumerate(self.options):
            print(f"{index + 1}. {option}")


class TextQuestion(Question):
    def __init__(self, description):
        super().__init__(description)

    def print_question_choices(self):
        print("Answer: _______________")


class RangeQuestion(Question):
    def __init__(self, description):
        super().__init__(description)

    def print_question_choices(self):
        print("Minimum: ______________")
        print("Maximum: ______________")


# this is open to be extended
# closed for modification as we do not need to touch this function
def print_quiz(questions):
    for question in questions:
        print(question.description)
        question.print_question_choices()


# only need to extend new features here
questions = [
    BooleanQuestion("This video is useful."),
    MultipleChoiceQuestion(
        "What is your favorite language?",
        [
            "CSS",
            "HTML",
            "JS",
            "Python",
        ],
    ),
    TextQuestion("Describe your favorite JS feature."),
    RangeQuestion("What is the speed limit in your city?"),
]

print_quiz(questions)
